File: dice.py
import re
from random import randint

diceRegex = re.compile("^(\d+)(?:d(\d*)?(?:([kdx])(\d+))?)?$")

def calc_throw(s):
    m = diceRegex.match(s)
    if not m:
        raise AttributeError("invalid string: %r"%s)
    throws, sides, typ, typ_count = m.groups()
    throws = int(throws)
    # we're done if it's a literal
    if sides is None:
        return throws
    sides = int(sides or 6) # default number of sides
    all_throws = [randint(1,sides) for _ in range(throws)]
    print("throws:", all_throws)
    if typ is None:
        return sum(all_throws)
    typ_count = int(typ_count)
    if typ == "d":
        assert typ_count < throws, "Invalid string, too many drops: %r"%s
        all_throws = sorted(all_throws)[typ_count:]
        return sum(all_throws)
    if typ == "k":
        assert typ_count < throws, "Invalid string, too many keeps: %r"%s
        all_throws = sorted(all_throws)[-typ_count:]
        return sum(all_throws)
    assert typ == "x", "Programming error this should NEVER happen"
    assert 0 < typ_count <= sides, "bad limit number: %r"%s
    all_throws = [x for x in all_throws if x < typ_count]
    while(len(all_throws) < throws):
        throw = randint(1,sides+1)
        if throw < typ_count:
            all_throws.append(throw)
            print("new throws: ", all_throws)
        else:
            print("bad throw", throw)
    return sum(all_throws)
    
    
for s in ["23", "3d", "2d7", "4d8k2", "4d8d1", "4d8x1"]:
    try:
        print (calc_throw(s))
    except Exception as e:
        print(e)

File: dice.js

// matches one roll spec
const diceRegex = /^(\d+)$|^(\d*)d(\d+)(?:([kdx])(\d+))?$/;
const operatorRegex = / *([-+]) */;

// rand int is 1-max inclusive e.g. 1 roll of n sided die
const randInt =  (max) =>  Math.ceil(Math.random() * max);
 
// multiple rolls of a die
function* sRolls(rolls, sides) {
    for  (let i=0; i < rolls; i++) {
        yield randInt(sides);
    }
}
//[...sRolls(5, 6)].sort().splice(2)  // => "5d6d2"

// "exploding" rolls of die if value is >= limit
function* eRolls(rolls, sides, limit) {
    for (let i=0; i < rolls;) {
      let roll = randInt(sides);
      yield roll;
      if (roll < limit) i++;
    }
}
//[...eRolls(5, 6, 3)]  // => "5d6e3"

var dieRoll = (expr) => {
  let match =  diceRegex.exec(expr);
  if (match)
    var [rolls, literal, N, sides, type, typeArg, ...rest] = match, tot;
  else
    throw("could not parse element:"+expr)
  // maybe just a literal
  if (literal)
    return {rolls: [], tot: parseInt(literal)}
  // ok we're going to roll ... at least once
  N = (N.length > 0) ? parseInt(N) : 1;
  sides = parseInt(sides);  // optional line??
  switch (type) {
  case undefined: //plain roll
    rolls = [...sRolls(N, sides)];
    break;
  case "d":  //drop
    assert(typeArg <= N, "Can't drop more than you have: "+expr);
    rolls = [...sRolls(N, sides)];
    rolls =  rolls.sort().slice(typeArg);
    break;
  case "k": //keep
    assert(typeArg <= N, "Can't keep more than you have: "+expr);
    rolls = [...sRolls(N, sides)];
    rolls = rolls.sort().slice(rolls.length-typeArg);
    break;
  case "x":  //explode
    assert(typeArg > 1, "this expolde arg will run forever: "+expr)
    rolls = [...eRolls(N, sides, typeArg)];
    break;
  default:
    throw ("you should not be able to get here but you did with: "+exp)
  }   
  tot =  rolls.reduce((a,b)=>a+b)
  return {rolls, tot}
}

const evalRoll = (expression) => {
  let uiArr = expression.split(/ *([-+]) */)
  // lets start with the first roll (or literal)
  let roll = dieRoll(uiArr[0]);
  let rolls = [roll.rolls];
  let tot = roll.tot;
  // keep rolling while there is more
  for (let i=1; i<uiArr.length;){
    let plusminus = uiArr[i++];
    roll = dieRoll(uiArr[i++]);
    rolls.push(roll.rolls);
    if (plusminus == "+")
      tot += roll.tot;
    else if (plusminus == "-")
      tot -= roll.tot;
    else
      throw("bad operator: "+plusminus);
  }
  return {rolls, tot}
}

module.exports = {
     dieRoll: dieRoll,
    evalRoll: evalRoll,
}

// test it out
evalRoll("2d9 + 3d6k2 +3d8d2- d12 +1000 - 5d4x4")
